prog/dftb+/lib_dftbplus/mainapi.F90

```diff
@@ -188,10 +191,433 @@ function nrOfAtoms()
+    call updateEquivalencyRelations(species0, iEqOrbitals, nIneqOrb, nMixElements)
+    call updateBLACSDecomposition(env, denseDesc)
+    call reallocateHSArrays(env, denseDesc, HSqrCplx, SSqrCplx, eigVecsCplx, HSqrReal, SSqrReal, eigVecsReal)
+    !If atomic order changes, partial charges need to be initialised, else wrong charge will be associated
+    !with each atom
+    call initialiseCharges(species0, speciesName, orb, nEl, iEqOrbitals, nIneqOrb, nMixElements,&
+         q0, qShell0, qInput, qInpRed, qOutput, qOutRed, qDiffRed)
+    
+  end subroutine updateDataDependentOnSpeciesOrdering
+
+  
+  !> Set shell-resolved partial charges.
+  !> If no argument is given, the final charges from the prior calculation are used 
+  subroutine setShellResolvedCharges(qSeed)
+    use dftbp_initprogram, only: qInput, qOutput, qInpRed, qOutRed, iEqOrbitals, nIneqOrb, orb 
+    use dftbp_spin,        only: qm2ud,ud2qm
+    use dftbp_orbitalequiv,only: OrbitalEquiv_reduce
+    
+    real(dp), allocatable, intent(in), optional :: qSeed(:, :, :)
+    
+    if(present(qSeed)) then
+       @:ASSERT(size(qSeed,1) == size(qInput,1))
+       @:ASSERT(size(qSeed,2) == size(qInput,2))
+       @:ASSERT(size(qSeed,3) == size(qInput,3))
+       qInput = qSeed
+       !Fill qInpRed
+       if (nSpin == 2) call qm2ud(qInput)
+       call OrbitalEquiv_reduce(qInput, iEqOrbitals, orb, qInpRed(1:nIneqOrb))
+       if (nSpin == 2) call ud2qm(qInput)
+    else
+       qInput  = qOutput
+       qInpRed = qOutRed 
+    endif
+    
+    qOutput = 0._dp
+    qOutRed = 0._dp
+ 
+  end subroutine setShellResolvedCharges
+
+   
 !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
 !!!  Private routines
 !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
 
+  
+  !> Update order of nr. atomic orbitals for each atom, orb%nOrbAtom
+  function updateAtomicOrbitals(species0) result(nOrbAtomReordered)
+    use dftbp_initprogram, only: orb, nAtom
+    !> Type of the atoms (nAtom)
+    integer,         allocatable,  intent(in) :: species0(:)
+    !> Nr. of orbitals for each atom (nAtom)  
+    integer,         allocatable :: nOrbAtomReordered(:)
+
+    @:ASSERT(nAtom == size(species0))
+    allocate(nOrbAtomReordered(nAtom))
+    nOrbAtomReordered(:) = orb%nOrbSpecies(species0(:)) 
+  end function updateAtomicOrbitals
+
+  
+  !> Update atomic masses
+  function updateAtomicMasses(species0) result(massReordered)
+    use dftbp_initprogram, only: speciesMass, nAtom
+    !> Type of the atoms (nAtom)
+    integer,  allocatable,   intent(in) :: species0(:)
+    !> List of atomic masses (nAtom)
+    real(dp), allocatable :: massReordered(:)
+
+    @:ASSERT(nAtom == size(speciesMass))
+    allocate(massReordered(size(speciesMass)))
+    massReordered = speciesMass(species0)
+  end function updateAtomicMasses
+
+  
+  ! TODO(Alex) This contains copy and pasted code from initprogram.F90
```

Comment:
Wrap code in this subroutine into smaller subroutines? Same code currently appears in initprogram.F90

Reply:
Yes, absolutely. One should refactor the code into a separate subroutine in `initprogram.F90` and call that. (Whenever you copy-paste, there is something wrong with the code structure :wink:)